fix($browser): normalize inputted URLs

jbedard opened the pull request:
<!-- General PR submission guidelines https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#submit-pr -->
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**
fix


**What is the current behavior? (You can also link to an open issue here)**



**What is the new behavior (if this is a feature change)?**



**Does this PR introduce a breaking change?**



**Please check if the PR fulfills these requirements**
- [ ] The commit message follows our [guidelines](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#commits)
- [ ] Fix/Feature: [Docs](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#documentation) have been added/updated
- [ ] Fix/Feature: Tests have been added; existing tests pass

**Other information**:

Narretz commented on src/ng/browser.js:
```diff
@@ -131,6 +131,15 @@ function Browser(window, document, $log, $sniffer) {
       var sameState = lastHistoryState === state;
+      var hasTrailingSlash = url[url.length - 1] === '/';
+
+      // Normalize the inputted URL
```
Should add an ellipsis at the end of this comment to match the one below

Narretz commented on src/ng/browser.js:
```diff
@@ -131,6 +131,15 @@ function Browser(window, document, $log, $sniffer) {
     // setter
     if (url) {
       var sameState = lastHistoryState === state;
+      var hasTrailingSlash = url[url.length - 1] === '/';
```
Regex would look nicer and could be re-used below

Narretz reviewed (commented):
Just a few minor things, but technically good to go

jbedard committed:
test($browser): update MockWindow to normalize URLs similar to real window.location

jbedard commented on src/ng/browser.js:
```diff
@@ -131,6 +131,15 @@ function Browser(window, document, $log, $sniffer) {
+      var hasTrailingSlash = url[url.length - 1] === '/';
+
+      // Normalize the inputted URL
+      url = urlResolve(url).href;
```
I think this is actually wrong, and I'm surprised it hasn't caused issues. This `url` should be the full url, that is protocol+domain+path+params+hash, basically equivalent to `window.location.href`...

jbedard committed:
fix($browser): normalize inputted URLs

Calls to `$browser.url` now normalize the inputted URL ensuring multiple
calls only differing in formatting do not force a browser `pushState`.

Normalization is done the same as the browser location URL and may
differ per browser and may be changed by browsers. Today no browsers
fully normalize URLs so this does not fix all instances of this issue.

See #16100
Closes #16606

jbedard commented:
Updated. Sorry I squashed/rebased everything so it might be hard to see the changes.

I added a commit (21263c9c546cf628bc25be450bd18079d3a66f18) refactoring the mock `window.location` in the `$browser` specs. I think this makes it more realistic by normalizing URLs assigned to `window.location.*` similar to the real `Location`. This was required for the tests added to the main commit to fail before (and pass after) the fix.

Also, as @petebacondarwin pointed out, none of the tests in the first commit actually fail in master right now. So I'm tempted to just drop them. I swear I saw these fail at some point though. @dmartres can you confirm your tests (from #16098) did fail at some point? Any idea why they are not be failing?

Narretz commented on src/ng/browser.js:
```diff
@@ -131,6 +131,16 @@ function Browser(window, document, $log, $sniffer) {
+
+      // ... but keep the (lack of) trailing '/' consistent
+      if (!hasTrailingSlash && trailingSlashRe.test(url)) {
+        url = url.replace(trailingSlashRe, "");
```
Should be single quote

Narretz commented on test/ng/browserSpecs.js:
```diff
@@ -567,6 +568,30 @@ describe('browser', function() {
+          browser.onUrlChange(callback);
+          browser.url('http://server/abc?q=\'');
+
+          browser.$$checkUrlChange();
```
Since we are testing a private API here, I feel like maybe we should test this in $location spec instead?

jbedard replied:
To avoid invoking the private API from tests? It really is testing logic only in browser.js though :/

gkalpak commented on src/ng/browser.js:
```diff
@@ -131,6 +131,16 @@ function Browser(window, document, $log, $sniffer) {
+      // Normalize the inputted URL ...
+      url = urlResolve(url).href;
+
+      // ... but keep the (lack of) trailing '/' consistent
```
Is it possible that there was a slash initially and `urlResolve` stripped it off?

jbedard replied:
I guess if a browser "normalizes" to have no slash at the end then yes, and I'd vote to handle that just in case. WDYT?

gkalpak replied:
It can't hurt. Let's do it :smiley:

jbedard replied:
Although now I'm wondering if trimming this is actually a problem :thinking:  (originally I didn't trim anything, but we added this to avoid adding `/` to many tests).

The whole point of this is to make the persisted `lastBrowserUrl` normalized the same way `document.location.href` is normalized. So the comparison below and maybe also the comparison in `$location` would not fail due to the normalized value being different then the non-normalized `lastBrowserUrl`.  By removing this `/` we are making `lastBrowserUrl` different then the normalized value...

Narretz replied:
But we are normalizing all the time, so it shouldn't matter? Or if it does, we should be able to add a test that shows this.

jbedard replied:
I think if we strip the slash here and assign to `lastBrowserUrl`, then later we compare to `$browser.url()` (which is `location.href`, which has a slash), then maybe it will detect changes when it shouldn't?
[browser.js#L148](https://github.com/jbedard/angular.js/blob/7467049dcd09efc5397213772a2c85a488bff1db/src/ng/browser.js#L148)
[browser.js#L231](https://github.com/jbedard/angular.js/blob/7467049dcd09efc5397213772a2c85a488bff1db/src/ng/browser.js#L231)

I won't change anything unless I can prove it with a test though...

jbedard replied:
I think we already have a test that proves it actually (but I'll try to add more).

Notice when I updated the `MockWindow` to use `aElement.href` to normalize the `window.location.href` mock... [one test](https://github.com/jbedard/angular.js/commit/21263c9c546cf628bc25be450bd18079d3a66f18#diff-c41da1c6661f19ace0ef7690f777fbcfR695) also had it's input changed otherwise it failed. I think the input shouldn't have to change and the fact that it failed shows what I'm talking about here... if we are comparing `lastBrowserURL` to `$browser.url()` then they must be normalized the same way. `$browser.url()` is `location.href`, so `lastBrowserURL` should be normalized just like `location.href`.

jbedard replied:
Yeah. I think we must keep the browser-normalized URL, slash or not.

See the new tests in the latest commit, 4 fail with the slash-removal logic (and they are run twice, so 8 failures total). Without the `urlResolve` there are 14 failures.

gkalpak commented on test/ng/browserSpecs.js:
```diff
@@ -567,6 +568,30 @@ describe('browser', function() {
         });
+
+        it('should not do pushState with a URL only different in encoding', function() {
+          //A URL from something such as window.location.href
```
Space after `//`.

jbedard commented:
Tests are currently failing in IE because IE is not normalizing the URLs like other browsers. Most likely the same IE bug that his block is solving: https://github.com/angular/angular.js/blob/v1.7.4/src/ng/urlUtils.js#L65-L71

gkalpak commented:
> Should we add specific handling for those chars the RFC lists? Or just respect what a.href = "..." does and have some potential browser inconsistencies?

Tough one :grin:
I vote for special handling if it only affects IE and browser inconsistencies otherwise.

jbedard commented:
I think this could get ugly :|

Chrome:
* encodes `'` in params
* decodes `-_.` in paths

IE:
* decodes `-_.` in paths

Firefox:
* encodes `'` in params
* decodes `.` in the paths

Everything else in `$-_.+!*'(),` seems to not be changed by `urlResolve`, so I believe this bug would be reproduced with those.

I'm going to sleep on it :|

petebacondarwin commented:
URL handling is hard

jbedard commented:
I've updated this to (I think) fix it better, although that makes the change bigger and potentially more questionable about merging vs wontfix.

This still adds URL normalizing to `$browser.url(newValue)` by using `urlResolve` like it originally did, but I've now updated `urlResolve` to normalize things that browsers don't do on their own such as decoding unnecessarily encoded things and making all hex-encoding upper case.

There's a whole [section in RFC3986](https://tools.ietf.org/html/rfc3986#section-6) talking about normalization/comparison and defines the ABNF for the [path](https://tools.ietf.org/html/rfc3986#section-3.3), [query](https://tools.ietf.org/html/rfc3986#section-3.4) and [fragment](https://tools.ietf.org/html/rfc3986#section-3.5). Those are the 3 parts of the URL this PR tries to normalize.

(Note that this [RFC3986](https://tools.ietf.org/html/rfc3986) replaces the one I was [previously referencing](https://github.com/angular/angular.js/pull/16606#issuecomment-427271178))

petebacondarwin reviewed (changes requested):
This looks very interesting. Can we get Travis green?

* Some linting issues https://travis-ci.org/angular/angular.js/jobs/441985047
* Lots of failed unit tests that look legit: https://travis-ci.org/angular/angular.js/jobs/441985048
* Possible failed e2e tests

jbedard commented:
But Chrome and FF did just fine :(

jbedard commented:
Most of the failures were because all browsers except Chrome+FF add the `?` and `#` if you assign to `HTMLAnchorElement.search/hash`. However now that I avoided that assignment when unnecessary... Chrome+FF no longer remove the `?` and `#` when empty.

Might require some manually trimming of `?#` :/

jbedard commented:
For reference: here's a commit which attempts to normalize encoding in the pathname/search/hash: https://github.com/jbedard/angular.js/commit/5d33cee9721b5e5a64cde7b4e2e100168a88bc11

While this normalizes a lot of things it also introduces some ugliness and inconsistencies across browsers (appending blank `?` and `#`), so for now I'm dropping this commit.

mgol reviewed (approved):
Looks good but it'd be great if you could add more details in the final commit message...
